src/Components/Endpoints/src/Binding/Converters/DictionaryConverter.cs

```diff
@@ -0,0 +1,81 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+namespace Microsoft.AspNetCore.Components.Endpoints.Binding;
+
+internal abstract class DictionaryConverter<TDictionary> : FormDataConverter<TDictionary>
+{
+}
+
+internal sealed class DictionaryConverter<TDictionary, TDictionaryPolicy, TBuffer, TKey, TValue> : DictionaryConverter<TDictionary>
+    where TKey : IParsable<TKey>
+    where TDictionaryPolicy : IDictionaryBufferAdapter<TDictionary, TBuffer, TKey, TValue>
+{
+    private readonly FormDataConverter<TValue> _valueConverter;
+
+    public DictionaryConverter(FormDataConverter<TValue> elementConverter)
+    {
+        ArgumentNullException.ThrowIfNull(elementConverter);
+
+        _valueConverter = elementConverter;
+    }
+
+    internal override bool TryRead(
+        ref FormDataReader context,
+        Type type,
+        FormDataMapperOptions options,
+        [NotNullWhen(true)] out TDictionary? result,
+        out bool found)
+    {
+        TValue currentValue;
+        TBuffer buffer;
+        bool foundCurrentValue;
+        bool currentElementSuccess;
+        bool succeded = true;
+
+        found = context.GetKeys().GetEnumerator().MoveNext();
+        if (!found)
+        {
+            result = default!;
+            return true;
+        }
```

Comment:
Probably not a big deal, but I wonder if we could avoid the extra enumerator allocation by instead setting `found` in the loop below and creating the buffer lazily.

Reply:
I've removed the allocation in a separate change that removes all allocations in this code-path where I implement the struct enumerator pattern.